sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -396,6 +396,16 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val REQUIRE_ALL_JOIN_KEYS_AS_PARTITION_KEYS =
+    buildConf("spark.sql.join.requireAllJoinKeysAsPartitionKeys")
```

Comment:
would this config take effect for all physical operators having 2 children and requiring `ClusteredDistribution`? Example like `CoGroupExec`.

Reply:
Yes, and they use `HashClusteredDistribution` before.

Reply:
Then should we make the config name more general?

Reply:
how about `spark.sql.requireAllClusterKeysAsPartitionKeysToCoParition`?

Reply:
SGTM!

Reply:
I'm slightly inclined to a shorter name like `spark.sql.join.requireAllJoinKeysAsPartitionKeys` but it's just a personal flavor. Also I'd suggest something like `spark.sql.enableStrictShuffleKeysCheck` but it's up to you.